File: Leetcode/Leetcode-DSA-course/Hashing/2473-max-sum-of-a-pair-with-equal-sum-of-digits/max-sum-of-a-pair-with-equal-sum-of-digits.py
class Solution:
    def maximumSum(self, nums: List[int]) -> int:
        def getDigitSum(n):
            sum = 0
            while n:
                sum += n % 10
                n //= 10
            return sum

        dic = defaultdict(int)
        ans = -1
        for num in nums:
            digitSum = getDigitSum(num)
            if digitSum in dic:
                ans = max(ans, dic[digitSum] + num)
            dic[digitSum] = max(dic[digitSum], num)
        return ans

File: Leetcode/Leetcode-DSA-course/Hashing/2473-max-sum-of-a-pair-with-equal-sum-of-digits/max-sum-of-a-pair-with-equal-sum-of-digits.cpp
class Solution {
public:
    int getDigitSum(int n) {
        int digitSum = 0;
        while (n) {
            digitSum += n % 10;
            n /= 10;
        }
        return digitSum;
    }

    int maximumSum(vector<int>& nums) {
        unordered_map<int, int> dic;
        int ans = -1;
        for (int num : nums) {
            int digitSum = getDigitSum(num);
            if (dic.find(digitSum) != dic.end()) {
                ans = max(ans, dic[digitSum] + num);
            }
            dic[digitSum] = max(dic[digitSum], num);
        }
        return ans;
    }
};